### PR TITLE
Feature/select patch

### PR DESCRIPTION
**Pull request by snudler6**

### CODE REVIEW

File: src/time_travel/patchers/select_patcher.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                break
+
+        if timeout is None and len(result_events) == 0:
+            raise ValueError('No relevant future events were set fot infinite '
+                             'timout')
+
+        self.events_pool.remove_events(result_timestamp, result_events)
+
+        self.clock.timestamp = result_timestamp
+        return result_events
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: You are not returning the correct `select` interface. `select` returns a tuple containing 3 list - ([r_ready], [w_ready], [somethinig_that_no_one_knows_or_uses]). You returned a list of all satisfied events, read or write altogether.

File: src/time_travel/patchers/select_patcher.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        result_timestamp = timeout_timestamp
+        
+        for timestamp, events in self.events_pool.get_events():
+            if timestamp > timeout_timestamp:
+                # No event before the timeout
+                break
+            
+            triggering_events = waited_events.intersection(events) 
+            
+            if len(triggering_events) != 0:
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: `if triggering_events:`

File: src/time_travel/patchers/select_patcher.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                result_events = list(triggering_events)
+                result_timestamp = timestamp
+                break
+
+        if timeout is None and len(result_events) == 0:
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: `if timeout is None and not result_events:`

File: src/time_travel/patchers/basic_patcher.py [hunk 4]
```diff
@@ -4,11 +4,13 @@
 class BasicPatcher(object):
     """Base class for patching time modules."""
     
-    def __init__(self, clock):
+    def __init__(self, clock, events_pool=None):
         """Create the patch."""
         self.clock = clock
-        self.patches = []
+        self.events_pool = events_pool
```
Comment (FarmerSez) [inline src/time_travel/patchers/basic_patcher.py]: Not sure this member belongs in this class. It obviously doesn't use it - why not just put it in the child class?

File: src/time_travel/events_pool.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+class EventsPool(object):
+    """Events set for handling patches of event drived libraries."""
+    
+    def __init__(self):
+        """Initialise an event pool."""
+        self.future_events = {}
+        
+    def add_future_event(self, timestamp, event):
+        """Add an event to a given timestamp."""
```
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: Is the timestamp absolute? is it from now? what is the unit of measurement?

File: src/time_travel/events_pool.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        self.future_events.setdefault(timestamp, set()).add(event)
+        
+    def get_events(self):
+        """Return all added events sorted by timestamp.
+        
+        The returned list is:
+        [(timestamp1, set(events)), (timestamp2, set(events)), ....]
+        where the timestamps are sorted.
+        """
+        return sorted(self.future_events.items())
```
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: I would explicitly give a `key=lambda x: x[0]` argument to `sorted`, but your call.

File: src/time_travel/events_pool.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    
+    def set_timestamp(self, timestamp):
+        """Remove all events before the given timestamp.
+        
+        This method should be called automatically after signing on the
```
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: This sentence is unclear. What does it mean? Is that something you can enforce by code?

File: src/time_travel/events_pool.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        """Return all added events sorted by timestamp.
+        
+        The returned list is:
+        [(timestamp1, set(events)), (timestamp2, set(events)), ....]
+        where the timestamps are sorted.
+        """
+        return sorted(self.future_events.items())
+    
+    def set_timestamp(self, timestamp):
+        """Remove all events before the given timestamp.
```
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: But the method name only specifies that you're setting the time. You should either rename the method name or give a better explanation to what's going on in it.

File: src/time_travel/events_pool.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        This method should be called automatically after signing on the
+        TimeMachineClock.
+        """
+        self.future_events = {k: v for k, v in self.future_events.items()
+                              if k >= timestamp}
+        
+    def remove_events(self, timestamp, events):
+        """Remove a list of events from a single timestamp."""
+        for event in events:
+            self.future_events.get(timestamp, set()).remove(event)
```
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: I'd definitely raise a ValueError if one of the events is not in it's set. That means that someone messed up his UT.

File: src/time_travel/patchers/tests/test_select.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        self.clock = TimeMachineClock(clock_listeners=[self.events_pool])
+        
+        self.patcher = SelectPatcher(self.clock, self.events_pool)
+        self.patcher.start()
+    
+    def teardown_method(self, method):
+        """Stop the select patcher"""
+        self.patcher.stop()
+
+    def test_empty_lists(self):
```
Comment (FarmerSez) [inline src/time_travel/patchers/tests/test_select.py]: Add a simple happy-flow case, these are all negative and edge-cases.

File: src/tests/test_time_travel.py [hunk 11]
```diff
@@ -37,3 +39,36 @@ def test_sleep_changing_today():
         
         time.sleep(3600)
         assert datetime.today() == datetime.fromtimestamp(3600)
+
+
+def test_select_no_timeout():
```
Comment (FarmerSez) [inline src/tests/test_time_travel.py]: Why are those 3 tests here and not in the select ut file?

File: src/time_travel/time_machine_clock.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+"""A single clock for all patchers."""
+
+
+class TimeMachineClock(object):
+    """Unifing class for clock types."""
+    
+    def __init__(self, start_timestamp=0.0, clock_listeners=[]):
```
Comment (snudler6) [inline src/time_travel/time_machine_clock.py]: Done

File: src/time_travel/patchers/tests/test_select.py (hunk 10 above)
Comment (snudler6) [inline src/time_travel/patchers/tests/test_select.py]: Done.

File: src/time_travel/events_pool.py (hunk 9 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Because of remove there will be a KeyError Raised which to me sounds more explanatory.

File: src/time_travel/events_pool.py (hunk 7 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Explained better in the doc.

File: src/time_travel/events_pool.py (hunk 8 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Explained better in the doc.

File: src/time_travel/events_pool.py (hunk 6 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Done.

File: src/time_travel/events_pool.py (hunk 5 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Done.

File: src/time_travel/patchers/basic_patcher.py (hunk 4 above)
Comment (snudler6) [inline src/time_travel/patchers/basic_patcher.py]: The idea is that all patchers can have access to the event_pool and the clock. There is no need to initialize it in all of them. (There will be more classes using the event pool)

File: src/time_travel/patchers/select_patcher.py (hunk 3 above)
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Done

File: src/time_travel/patchers/select_patcher.py (hunk 2 above)
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Done

File: src/time_travel/time_machine_clock.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        self._timestamp = start_timestamp
+        self._clock_listeners = clock_listeners
+        
+    @property
+    def timestamp(self):
+        """Get the clock timestamp in seconds since the epoch."""
+        return self._timestamp
+    
+    @timestamp.setter
+    def timestamp(self, timestatmp):
```
Comment (snudler6) [inline src/time_travel/time_machine_clock.py]: Done.

File: src/time_travel/time_machine_clock.py (hunk 12 above)
Comment (snudler6) [inline src/time_travel/time_machine_clock.py]: Done.

File: src/tests/test_time_travel.py (hunk 11 above)
Comment (snudler6) [inline src/tests/test_time_travel.py]: I do want some tests with select integration with other patchers.
They test things tested in the unit-test but integrated with the other patchers

File: src/time_travel/patchers/select_patcher.py (hunk 1 above)
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Thanks!
Done

File: src/time_travel/time_travel.py [hunk 14]
```diff
@@ -1,65 +1,38 @@
...
-print sys.path
 from patchers.datetime_patcher import DatetimePatcher
 from patchers.time_patcher import TimePatcher
+from patchers.select_patcher import SelectPatcher
 
+from time_machine_clock import TimeMachineClock
+from events_pool import EventsPool
+
+import sys
+print sys.path
```
Comment (FarmerSez) [inline src/time_travel/time_travel.py]: Residue.

File: src/time_travel/time_machine_clock.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
+"""A single clock for all patchers."""
+
+
+class TimeMachineClock(object):
```
Comment (FarmerSez) [inline src/time_travel/time_machine_clock.py]: Come to think of it, I'm not sure the appropriate term in this class is `timestamp`. A timestamp is the time in which something has happened. I think that it would be better to use `clock.time += 5`, and not `clock.timestamp += 5`.

It might also be handy to create another class for the time itself, that way you could override the `=`, `+=`, `-=` operators.

File: src/time_travel/patchers/tests/test_select.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        assert select.select([event], [], []) == ([event], [], [])
+        assert self.clock.timestamp == 3
+
+    def test_select_infinite_wait(self):
+        event = mock.MagicMock()
+        
+        with pytest.raises(ValueError):
+            select.select([event], [], [])
+            
+    def test_event_returned_in_multiple_lists(self):
```
Comment (FarmerSez) [inline src/time_travel/patchers/tests/test_select.py]: That is not a good test, and it raises an issue. You have to define in your event pool event types. 
Usually when you have a single socket in both the r_list and w_list is when you buffer data in an incoming and outgoing FIFOs. In that case, if the OUT_FIFO has data and the socket is write ready - you'd read data from the fifo and send it to the socket. If the socket is read ready, you'd read from it and put the data in the IN_FIFO.

If you don't allow users to chose which event has happened in their socket, they wouldn't be able to test specific cases in their UT.

File: src/time_travel/patchers/select_patcher.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+        if timeout is None and not result_events:
+            raise ValueError('No relevant future events were set fot infinite '
+                             'timout')
+
+        self.events_pool.remove_events(result_timestamp, result_events)
+
+        self.clock.timestamp = result_timestamp
+    
+        return (self._list_intersection(result_events, rlist),
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: As I explained below, this is a really misleading behaviour. You have to add an event type to the event pool.

One more thing that bugs me is that you referred to the given and returned object from select as `events`. This is incorrect - `select` receives a `selectable` object, and defines the events that occurred by the list that this object is returned at.

I would change the interface to `pool.add_event(selectable, pool.READ_EVENT, timeout=6)` or to `pool.add_read_event(selectable, timeout=6)`. Something like that.
The former will be handy, because if you're planning on mocking `epoll` - it only works with event masks, so the interface will be similar.

File: src/time_travel/events_pool.py (hunk 9 above)
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: Gotcha, somehow missed that it raises an exception if there's no event in the set. However, if there are no events in that timestamp whatsoever, the exception would be that the set is empty, rather than there were no events at that time to begin with. Shouldn't you check that separately? The more explicit you get here the better.

Also, if the set is empty after removal, shouldn't you delete the dictionary item for that timestamp?

File: src/time_travel/patchers/select_patcher.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        self.patches = [mock.patch('select.select', self.select)]
+        
+    @staticmethod
+    def _list_intersection(list1, list2):
+        return list(set(list1).intersection(set(list2)))
+        
+    def _mocked_select(self, rlist, wlist, xlist, timeout=None):
+        waited_events = set(rlist + wlist + xlist)
+        
+        added_timeout = timeout
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: I prefer one-liner ifs in that case: `added_timeout = float('inf') if timeout is None else timeout`. Only if it's below 80 cols though :^)

Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Done.

File: src/time_travel/events_pool.py (hunk 9 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Cool.
So, I'll return to what you offered in the beginning. trying to excess the timestamp will raise an error if there is no events on that timestamp.
removing the timestamp completely is basically for performance only, and it will be removed when the clock will pass over it, but I added it anyways.

File: src/time_travel/patchers/tests/test_select.py (hunk 16 above)
Comment (snudler6) [inline src/time_travel/patchers/tests/test_select.py]: Done.

File: src/time_travel/time_machine_clock.py (hunk 15 above)
Comment (snudler6) [inline src/time_travel/time_machine_clock.py]: Separated between time and timstamp.
Don't see any reason why float isn't good enough and to prove it I removed the advance_timestamp method.

File: src/time_travel/patchers/select_patcher.py (hunk 17 above)
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Done.

File: src/time_travel/events_pool.py [hunk 19]
```diff
@@ -8,23 +8,35 @@ def __init__(self):
...
         The returned list is:
         [(timestamp1, set(events)), (timestamp2, set(events)), ....]
         where the timestamps are sorted.
         """
-        return sorted(self.future_events.items(), key=lambda x: x[0])
+        def filtered_set(events_set):
+            return set([event for event, event_type in
+                        events_set if predicate(event_type)])
+        
+        predicate = (lambda etype: True) if predicate is None else predicate
```
Comment (FarmerSez) [inline src/time_travel/events_pool.py]: It took me a while to get where this is used. You should move it to be before the private function definition.

File: src/time_travel/patchers/basic_patcher.py [hunk 20]
```diff
@@ -10,6 +10,11 @@ def __init__(self, clock, events_pool=None):
         self.events_pool = events_pool
         
         self.patches = []
+        
+    @classmethod
+    def register_events_types(cls, main_events_types):
```
Comment (FarmerSez) [inline src/time_travel/patchers/basic_patcher.py]: This isn't a patcher, why does it have this interface?

File: src/time_travel/patchers/select_patcher.py [hunk 21]
```diff
@@ -13,6 +14,8 @@ class SelectPatcher(BasicPatcher):
         - select.select
     """
     
+    EventTypes = Enum('select', 'READ WRITE EXCEPTIONAL')
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: This isn't a type, it's an object. PEP8.

File: src/time_travel/patchers/select_patcher.py [hunk 22]
```diff
@@ -23,42 +26,74 @@ def __init__(self, *args, **kwargs):
         
         self.patches = [mock.patch('select.select', self.select)]
         
+    @classmethod
+    def register_events_types(cls, main_events_types):
+        """Register select events types."""
+        setattr(main_events_types, 'select', cls.EventTypes)
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: I think it's a bit crooked. Why does the patcher need to know how to set things in the library? You're gonna use entry points, that means that the patcher should tell the library how to register itself to it, not the other way around.
The patchers should expose an interface that will return the event types, and the main library needs to do all the registration business.

File: src/time_travel/patchers/select_patcher.py (hunk 21 above)
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: It's actually is a type:
https://docs.python.org/3/library/enum.html#functional-api

File: src/time_travel/events_pool.py [hunk 23]
```diff
@@ -8,23 +8,35 @@ def __init__(self):
         """Initialise an event pool."""
         self.future_events = {}
         
-    def add_future_event(self, timestamp, event):
+    def add_future_event(self, timestamp,
+                         event_descriptor,
+                         event_predicate=None):
```
Comment (snudler6) [inline src/time_travel/events_pool.py]: With poll with no mask?

File: src/time_travel/patchers/select_patcher.py (hunk 22 above)
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Done.

File: src/time_travel/patchers/select_patcher.py [hunk 24]
```diff
@@ -23,42 +26,74 @@ def __init__(self, *args, **kwargs):
...
                 break
-
-        if timeout is None and not result_events:
+            
+        return result_timestamp, result_events
+    
+    def _mocked_select(self, rlist, wlist, xlist, timeout=None):
+        read_timestamp, read_events = self._get_earliest_events(
+            rlist,
+            SelectPatcher.EventTypes.READ,
```
Comment (snudler6) [inline src/time_travel/patchers/select_patcher.py]: Done.

File: src/time_travel/events_pool.py [hunk 25]
```diff
@@ -8,23 +8,35 @@ def __init__(self):
...
         """Add an event to a given timestamp.
         
         - timestamp: time in seconds since the epoch.
```
Comment (snudler6) [inline src/time_travel/events_pool.py]: Done.

File: src/time_travel/events_pool.py (hunk 19 above)
Comment (snudler6) [inline src/time_travel/events_pool.py]: Done.

File: src/time_travel/patchers/basic_patcher.py (hunk 20 above)
Comment (snudler6) [inline src/time_travel/patchers/basic_patcher.py]: It's the virtual patcher. It doesn't any more, but it has the new interface to register patchers.

File: src/time_travel/time_machine_clock.py (hunk 15 above)
Comment (FarmerSez) [inline src/time_travel/time_machine_clock.py]: It's A-Okay, I just didn't know that having both setter and getter makes `+=` work on it :)
